.circleci/config.yml

```diff
@@ -0,0 +1,314 @@
+  run-test:
+    description: "Run a group of tests, ie 'formatting'."
+    parameters:
+      for:
+        type: string
+        default: all
+    steps:
+      - checkout
+      - restore_cache:
+          keys:
+            - v1-with-test-eggs-cache-{{ checksum "requirements.txt" }}-{{ checksum "test_requirements.txt" }}
+            - v1-cache-{{ checksum "requirements.txt" }}-{{ checksum "test_requirements.txt" }}
+      - run:
+          name: Test
+          command: |
+            source ~/code/venv/bin/activate
+            python setup.py test<< parameters.for >>
+      - run:
+          name: Upload Test Coverage
+          command: bash <(curl -s https://codecov.io/bash)
+      - save_cache:
+          key: v1-with-test-eggs-cache-{{ checksum "requirements.txt" }}-{{ checksum "test_requirements.txt" }}
+          paths:
+            - ~/code/.eggs
+            - ~/code/venv
+
+jobs:
+
+  # Make the docker images this project uses.
+  make-images:
+    machine:
+      docker_layer_caching: true
+    environment:
+      GORDO_PROD_MODE: true
+    steps:
+      - checkout
+      - run:
+          name: Make Images
+          command: make images
+
+  # Push images, only when on master and/or tagged commits.
+  push-images:
+    machine:
+      docker_layer_caching: true
+    environment:
+      GORDO_PROD_MODE: true
+    steps:
+      - checkout
+      - run:
+          name: Push Images
+          command: make images push-prod-images
+
+  # Publish to PyPi, only on tagged commits.
+  publish-to-pypi:
+    docker:
+      - image: circleci/python:3.6
+    steps:
+      - checkout
+      - restore_cache:
+          keys:
+            - v1-with-test-eggs-cache-{{ checksum "requirements.txt" }}-{{ checksum "test_requirements.txt" }}
+            - v1-cache-{{ checksum "requirements.txt" }}-{{ checksum "test_requirements.txt" }}
+      - run:
+          name: Build
+          command: |
+            source ~/code/venv/bin/activate
+            python setup.py sdist
+            python setup.py bdist_wheel
+      - run:
+          # Requires TWINE_PASSWORD to be set.
+          name: Publish
+          command: |
+            source ~/code/venv/bin/activate
+            pip install twine
+            twine upload -u travis-statoil --skip-existing dist/*
+
+
+  # Make the Sphinx documentation for the project; without errors or warnings.
+  make-docs:
+    docker:
+      - image: circleci/python:3.6
+    steps:
+      - checkout
+      - restore_cache:
+          keys:
+            - python-docs-cache-{{ checksum "requirements.txt" }}-{{ checksum "dev_requirements.txt"}}
+      - run:
+          name: Install Dependencies
+          command: |
+            python -m venv ~/venv
+            source ~/venv/bin/activate
+            pip install -r requirements.txt
+            pip install -r dev_requirements.txt
+      - save_cache:
+          key: python-docs-cache-{{ checksum "requirements.txt" }}-{{ checksum "dev_requirements.txt"}}
```

Comment:
I don't know if it will matter, but you have several caches which it seems like should be possible to optimize a bit:
1.  `python-docs-cache` with ` ~/venv`
2. `v1-with-test-eggs-cache` with `~/code/venv` and `~/code/.eggs`
3. `v1-cache` with `~/code/venv`
It probably possible to restore 2 caches, so then point 2 above can be converted to just `~/code/.eggs` maybe? And can `~/code/venv` and ` ~/venv` be the same?